src/FESpaces/DiscreteModelWithFEMaps.jl

```diff
@@ -0,0 +1,156 @@
+  grid::Grid{Dc,Dp}
+  fe_sp::A
+  scal_fe_sp::B
+  fe_map::C
+  node_coords::D
+  reffes::E
+  cell_type::F
+end
+
+function GridWithFEMap(model,reffe::ReferenceFE; kwargs...)
+  cell_type = Fill(1,num_cells(model))
+  _reffes = [reffe]
+  reffes = lazy_map(Reindex(_reffes),cell_type)
+  GridWithFEMap(model,cell_type,reffes; kwargs...)
+end
+
+function _cell_vector_to_dof_vector!(dof_vector,cell_node_ids, cell_vector)
+  cache_cell_node_ids = array_cache(cell_node_ids)
+  cache_cell_vector   = array_cache(cell_vector)
+  for k=1:length(cell_node_ids)
+    current_node_ids = getindex!(cache_cell_node_ids,cell_node_ids,k)
+    current_values   = getindex!(cache_cell_vector,cell_vector,k)
+    for (i,id) in enumerate(current_node_ids)
+      dof_vector[current_node_ids[i]]=current_values[i]
+    end
+  end
+end
+
+function GridWithFEMap(model,cell_type,reffes::AbstractArray{<:ReferenceFE}; kwargs...)
+
+  # Create a FESpace for the geometrical description
+  # The FEFunction that describes the coordinate field
+  # or displacement can be a interpolation or a solution
+  # of a mesh displacement problem
+  T = eltype(get_node_coordinates(model))
+
+  Vₕ = FESpace(model,reffes;conformity=:H1,kwargs...)
+
+  Ts = eltype(T)
+  os = get_order.(reffes)
+  ps = get_polytope.(reffes)
+  f(a,b) = LagrangianRefFE(Float64,a,b)
+  s_reffes = lazy_map(f,ps,os)
+  Vₕ_scal = FESpace(model,s_reffes;conformity=:H1)
+
+  grid = get_grid(model)
+  geo_map = get_cell_map(grid)
+
+  cell_ctype = get_cell_type(grid)
+  c_reffes = get_reffes(grid)
+
+  # Create a fe_map using the cell_map that can be evaluated at the
+  # vertices of the fe space (nodal type)
+  # This returns a FEFunction initialised with the coordinates
+  # But this is to build the FEFunction that will be inserted, it is
+  # an advanced constructor, not needed at this stage
+
+  c_dofs = get_fe_dof_basis(Vₕ)
+  dof_basis = get_data(c_dofs)
+  c_nodes = lazy_map(get_nodes,get_data(c_dofs))
+
+  xh = zero(Vₕ)
+  nc = get_node_coordinates(model)
+  c_dofv = lazy_map(evaluate,dof_basis,geo_map)
+
+  Uₕ = TrialFESpace(Vₕ)
+
+  fv = get_free_dof_values(xh)
+  dv = get_dirichlet_dof_values(get_fe_space(xh))
+  gather_free_and_dirichlet_values!(fv,dv,Uₕ,c_dofv)
+
+  c_xh = lazy_map(evaluate,get_data(xh),c_nodes)
+  c_scal_ids = get_cell_dof_ids(Vₕ_scal)
+
+
+  nodes_coords = Vector{eltype(eltype(c_xh))}(undef,num_free_dofs(Vₕ_scal))
+  _cell_vector_to_dof_vector!(nodes_coords,c_scal_ids,c_xh)
+
+  GridWithFEMap(grid, Vₕ, Vₕ_scal, xh, nodes_coords, reffes, cell_type)
+end
+
+function _compute_node_coordinates(grid,xh)
+  c_dofs = get_fe_dof_basis(grid.fe_sp)
+  c_nodes = lazy_map(get_nodes,get_data(c_dofs))
+
+  c_xh = lazy_map(evaluate,get_data(xh),c_nodes)
+  c_scal_ids = get_cell_dof_ids(grid.scal_fe_sp)
+
+  nodes_coords = grid.node_coords
+  _cell_vector_to_dof_vector!(nodes_coords,c_scal_ids,c_xh)
+
+  return nodes_coords
+end
+
+function add_mesh_displacement!(grid::GridWithFEMap,dh::FEFunction)
```

Comment:
To-think: is it possible to factor out `add_mesh_displacement!` and `update_coordinates!` with a third parameter of type `Function` in order to avoid code replication here?